c++/src/capnp/rpc.c++

```diff
@@ -1418,38 +1418,81 @@ private:
-      return rpcInner.writeDescriptor(descriptor, fds);
-    } else {
-      KJ_IF_SOME(exportId, exportsByCap.find(inner)) {
-        // We've already seen and exported this capability before.  Just up the refcount.
-        auto& exp = KJ_ASSERT_NONNULL(exports.find(exportId));
-        ++exp.refcount;
-        if (exp.resolveOp == kj::none) {
-          descriptor.setSenderHosted(exportId);
-        } else {
-          descriptor.setSenderPromise(exportId);
-        }
-        return exportId;
-      } else {
-        // This is the first time we've seen this capability.
-        ExportId exportId;
-        auto& exp = exports.next(exportId);
-        exportsByCap.insert(inner, exportId);
-        exp.canonical = true;
-        exp.refcount = 1;
-        exp.clientHook = inner->addRef();
-
-        KJ_IF_SOME(wrapped, inner->whenMoreResolved()) {
-          // This is a promise.  Arrange for the `Resolve` message to be sent later.
-          exp.resolveOp = resolveExportedPromise(exportId, kj::mv(wrapped));
-          descriptor.setSenderPromise(exportId);
-        } else {
-          descriptor.setSenderHosted(exportId);
+    KJ_IF_SOME(rpcInner, unwrapIfSameNetwork(*inner)) {
+      if (rpcInner.isOnConnection(*this)) {
+        return rpcInner.writeDescriptor(descriptor, fds);
+      }
+
+      KJ_IF_SOME(conn, tryGetConnection()) {
+        KJ_IF_SOME(otherConn, rpcInner.connectionState->tryGetConnection()) {
+          if (conn.canIntroduceTo(otherConn)) {
+            // We can do three-party handoff!
+
+            auto tph = descriptor.initThirdPartyHosted();
+
+            QuestionId questionId;
+            auto& question = rpcInner.connectionState->questions.nextHigh(questionId);
+            question.isAwaitingReturn = false;  // No Return needed
+            auto questionRef = kj::refcounted<QuestionRef>(
+                *rpcInner.connectionState, questionId, kj::none);
+            question.selfRef = *questionRef;
+
+            auto otherMsg = otherConn.newOutgoingMessage(messageSizeHint<rpc::Provide>() + 32);
+            auto provide = otherMsg->getBody().initAs<rpc::Message>().initProvide();
+            provide.setQuestionId(questionId);
+            rpcInner.writeTarget(provide.initTarget());
+            conn.introduceTo(otherConn, tph.initId(), provide.initRecipient());
+            otherMsg->send();
+
+            // We always have to add a new export to the table to use as the "vine"; we cannot
+            // share an existing export pointing to the same capability. The reason is, the vine
+            // represents not just the capability but *this specific handoff* of the capability;
+            // when it is dropped, we can clean up the handoff. This is accomplished by attaching
+            // the `questionRef` to the export, below.
+            //
+            // Additionally, the vine is never treated as a promise, even if the capability backing
+            // it is a promise.
```

Comment:
So if the vat doesn't understand the handoff, do they just get a half-finished capability forever? Even if the promise resolves to an error?